### DISCUSSION TITLE
Q: why not use F# as base language instead of a custom ML dialect?

### DISCUSSION THREAD

**Discussion by jkone27** [darklang/dark#4670] [Q&A] [answered]
Question: using F# as a base language instead of a custom ML dialect might make adoption easier? or what are the advantages of using a custom dialect? Thanks for this amazing project!

**Comment by pbiggar** [comment, marked as answer]
Most languages are designed for the kind of offline development that's common: they have compilers and build systems and package managers that download libraries onto a filesystem, and to deploy them you bundle them with their dependencies, webserver and a runtime into a container and put then on a cloud server somehow and connect to a DB. So the path from typing a keystroke is very very far from being in production.

So that means in practice that all languages are designed for a specific niche, and no language was designed for the darklang niche (trace-driven/deployless web backends). That niche means we need features like:
- concept of Incomplete
- concept of blanks / AST holes
- feature flags as part of the language
- impossible to have syntax errors
- no filesystem / OS/ container
- instant "compile"
- instrumented runtime
- language runtime integrated into DB, queues and web server

So F# for example was designed for the niche of a general purpose functional language in the .NET ecosystem. Which means it has some great features such as C#/CIL integration, but brings with it a build system, package system, and runtime that make it inappropriate for our niche.

A similar but targeted argument could be made about Python, Rails, Node, Go, Rust, etc.

Now we could just do something like pretend it's F#, by making the system closer and renaming some libraries. But then people would wonder why their existing code doesn't work, why they can't reuse libraries, why they can't integrate with C# code, etc.

**Reply by jkone27** [reply]
https://roadmap.darklang.com/editor/canvas.html btw this looks super cool,  i see your point was thinking about maybe to reduce developer learning-curve / friction with what they already know or already documented languages/grammar, but i understand your points. one notable example is e.g. Fable javascript, where you can still use "a (big) portion of the F# language" but you will actually run javascript on node. Maybe a Fable to Darklang compiler could be something one day, i also imagine developers more likely to develop maybe in vscode with an extension instead of directly in the browser but ofc i also understand the concept of "bringing everything in one place only", just maybe it could be a vscode based web editor? (vscode server kind of?) with an extension. just ideas
